kotlinx-coroutines-test/common/src/TestScope.kt

```diff
@@ -0,0 +1,219 @@
+    var scope: TestScopeImpl? = null
+    val exceptionHandler = when (ctxWithDispatcher[CoroutineExceptionHandler]) {
+        null -> CoroutineExceptionHandler { _, exception ->
+            scope!!.reportException(exception)
+        }
+        else -> throw IllegalArgumentException(
+            "A CoroutineExceptionHandler was passed to TestScope. " +
+                "Please pass it as an argument to a `launch` or `async` block on an already-created scope " +
+                "if uncaught exceptions require special treatment."
+        )
+    }
+    return TestScopeImpl(ctxWithDispatcher + exceptionHandler).also { scope = it }
+}
+
+/**
+ * Adds a [TestDispatcher] and a [TestCoroutineScheduler] to the context if there aren't any already.
+ *
+ * @throws IllegalArgumentException if both a [TestCoroutineScheduler] and a [TestDispatcher] are passed.
+ * @throws IllegalArgumentException if a [ContinuationInterceptor] is passed that is not a [TestDispatcher].
+ */
+internal fun CoroutineContext.withDelaySkipping(): CoroutineContext {
+    val dispatcher: TestDispatcher = when (val dispatcher = get(ContinuationInterceptor)) {
+        is TestDispatcher -> {
+            val ctxScheduler = get(TestCoroutineScheduler)
+            if (ctxScheduler != null) {
+                require(dispatcher.scheduler === ctxScheduler) {
+                    "Both a TestCoroutineScheduler $ctxScheduler and TestDispatcher $dispatcher linked to " +
+                        "another scheduler were passed."
+                }
+            }
+            dispatcher
+        }
+        null -> StandardTestDispatcher(get(TestCoroutineScheduler))
+        else -> throw IllegalArgumentException("Dispatcher must implement TestDispatcher: $dispatcher")
+    }
+    return this + dispatcher + dispatcher.scheduler
+}
+
+internal class TestScopeImpl(context: CoroutineContext) :
+    AbstractCoroutine<Unit>(context, initParentJob = true, active = true), TestScope {
+
+    override val testScheduler get() = context[TestCoroutineScheduler]!!
+
+    var entered = false
+    var finished = false
+    val uncaughtExceptions = mutableListOf<Throwable>()
+    val lock = SynchronizedObject()
+
+    /** Called upon entry to [runTest]. Will throw if called more than once. */
+    fun enter() {
+        val exceptions = synchronized(lock) {
+            if (entered)
+                throw IllegalStateException("Only a single call to `runTest` can be performed during one test.")
+            entered = true
+            check(!finished)
+            uncaughtExceptions
+        }
+        if (exceptions.isNotEmpty()) {
+            throw UncaughtExceptionsBeforeTest().apply {
+                for (e in exceptions)
+                    addSuppressed(e)
+            }
+        }
+    }
+
+    /** Called at the end of the test. May only be called once. */
+    fun finish(): List<Throwable> {
+        val exceptions = synchronized(lock) {
+            check(entered && !finished)
```

Comment:
If all these `check` calls can be triggered, then a (potentially actionable) diagnostic message should be added. Otherwise, it should be an assert

Reply:
`assert` is internal. Is it worth it to make it `PublishedApi`?

Reply:
Changed this.